File: packages-gen/dart/lang/lib/common-sort-by.dart
import 'package:xtalk_lang/common-data.dart' as xt.lang.common_data;

sort_by(arr, inputs) {
  var keys = xt.lang.common_data.arr_map(inputs,(e) {
    return ((e.runtimeType).toString().startsWith("List") || (e.runtimeType).toString().startsWith("_GrowableList")) ? e[0] : e;
  });
  var inverts = xt.lang.common_data.arr_map(inputs,(e) {
    return ((e.runtimeType).toString().startsWith("List") || (e.runtimeType).toString().startsWith("_GrowableList")) ? e[1] : false;
  });
  var get_fn = (e, key) {
    if((key.runtimeType).toString().contains("Function") || (key.runtimeType).toString().contains("=>") || (key).toString().startsWith("Closure")){
      return key(e);
    }
    else{
      return e[key];
    }
  };
  var key_fn = (e) {
    return xt.lang.common_data.arr_map(keys,(key) {
      return Function.apply((get_fn as Function),<dynamic>[e,key]);
    });
  };
  var comp_fn = (a0, a1) {
    var arr_39640 = a0;
    for(var i = 0; i < arr_39640.length; ++i){
      var v0 = arr_39640[i];
      var v1 = a1[i];
      var invert = inverts[i];
      if(v0 != v1){
        if((null != invert) && (false != invert)){
          if(("int" == (v0.runtimeType).toString()) || ("double" == (v0.runtimeType).toString()) || ("num" == (v0.runtimeType).toString())){
            return (("String" == (v1.runtimeType).toString()) || ("String" == (v0.runtimeType).toString())) ? ((v1).toString().compareTo((v0).toString()) < 0) : (v1 < v0);
          }
          else{
            return ((v1).toString()).toString().compareTo(((v0).toString()).toString()) < 0;
          }
        }
        else{
          if(("int" == (v0.runtimeType).toString()) || ("double" == (v0.runtimeType).toString()) || ("num" == (v0.runtimeType).toString())){
            return (("String" == (v0.runtimeType).toString()) || ("String" == (v1.runtimeType).toString())) ? ((v0).toString().compareTo((v1).toString()) < 0) : (v0 < v1);
          }
          else{
            return ((v0).toString()).toString().compareTo(((v1).toString()).toString()) < 0;
          }
        }
      }
    };
    return false;
  };
  var out = xt.lang.common_data.arr_clone(arr);
  xt.lang.common_data.arr_sort(out,key_fn,comp_fn);
  return out;
}

File: packages-gen/js/lang/common-sort-by.js
const xt.lang.common_data = require("@xtalk/lang/common-data.js")

function sort_by(arr,inputs){
  let keys = inputs.map(function (e){
    return Array.isArray(e) ? e[0] : e;
  });
  let inverts = inputs.map(function (e){
    return Array.isArray(e) ? e[1] : false;
  });
  let get_fn = function (e,key){
    if("function" == (typeof key)){
      return key(e);
    }
    else{
      return e[key];
    }
  };
  let key_fn = function (e){
    return keys.map(function (key){
      return get_fn(e,key);
    });
  };
  let comp_fn = function (a0,a1){
    for(let i = 0; i < a0.length; ++i){
      let v0 = a0[i];
      let v1 = a1[i];
      let invert = inverts[i];
      if(v0 != v1){
        if(invert){
          if("number" == (typeof v0)){
            return v1 < v0;
          }
          else{
            return 0 > String(v1).localeCompare(String(v0));
          }
        }
        else{
          if("number" == (typeof v0)){
            return v0 < v1;
          }
          else{
            return 0 > String(v0).localeCompare(String(v1));
          }
        }
      }
    };
    return false;
  };
  let out = arr.slice();
  xt.lang.common_data.arr_sort(out,key_fn,comp_fn);
  return out;
}

module.exports = {["sort_by"]:sort_by}